File: Easy/225. Implement Stack using Queues/225. Implement Stack using Queues.ts
class MyStack {
    private q1: number[];
    private q2: number[];

    constructor() {
        this.q1 = [];
        this.q2 = [];
    }

    push(x: number): void {
        this.q1.push(x);
    }

    pop(): number {
        while (this.q1.length > 1) {
            this.q2.push(this.q1.shift()!);
        }
        const result = this.q1.shift()!;
        [this.q1, this.q2] = [this.q2, this.q1];
        return result;
    }

    top(): number {
        return this.q1[this.q1.length - 1];
    }

    empty(): boolean {
        return this.q1.length === 0;
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: Easy/225. Implement Stack using Queues/225. Implement Stack using Queues.rs
use std::collections::VecDeque;

struct MyStack {
    q1: VecDeque<i32>,
    q2: VecDeque<i32>,
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyStack {

    fn new() -> Self {
        MyStack {
            q1: VecDeque::new(),
            q2: VecDeque::new(),
        }
    }
    
    fn push(&mut self, x: i32) {
        self.q1.push_back(x);
    }
    
    fn pop(&mut self) -> i32 {
        while self.q1.len() > 1 {
            self.q2.push_back(self.q1.pop_front().unwrap());
        }
        let result = self.q1.pop_front().unwrap();
        std::mem::swap(&mut self.q1, &mut self.q2);
        result
    }
    
    fn top(&self) -> i32 {
        self.q1.back().unwrap().clone()
    }
    
    fn empty(&self) -> bool {
        self.q1.is_empty()
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * let obj = MyStack::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: bool = obj.empty();
 */